Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalPlateManager.cpp

```diff
@@ -0,0 +1,166 @@
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "plMetalPlateManager.h"
+
+#include <assert.h>
+
+#include "ShaderTypes.h"
+#include "plMetalPipeline.h"
+
+plMetalPlateManager::plMetalPlateManager(plMetalPipeline *pipe)
+    : plPlateManager(pipe),
+      fVtxBuffer(nullptr)
+{
+    MTL::DepthStencilDescriptor *depthDescriptor = MTL::DepthStencilDescriptor::alloc()->init();
+    depthDescriptor->setDepthCompareFunction(MTL::CompareFunctionAlways);
+    depthDescriptor->setDepthWriteEnabled(false);
+    fDepthState = pipe->fDevice.fMetalDevice->newDepthStencilState(depthDescriptor);
+    depthDescriptor->release();
+}
+
+void plMetalPlateManager::ICreateGeometry()
+{
+    plMetalPipeline *pipeline = (plMetalPipeline *)fOwner;
+    if (!fVtxBuffer) {
+        struct plateVertexBuffer vertexBuffer;
+        vertexBuffer.vertices[0].Set(-0.5f, -0.5f);
+        vertexBuffer.uv[0].Set(0.0f, 0.0f);
+
+        vertexBuffer.vertices[1].Set(-0.5f, 0.5f);
+        vertexBuffer.uv[1].Set(0.0f, 1.0f);
+
+        vertexBuffer.vertices[2].Set(0.5f, -0.5f);
+        vertexBuffer.uv[2].Set(1.0f, 0.0f);
+
+        vertexBuffer.vertices[3].Set(0.5f, 0.5f);
+        vertexBuffer.uv[3].Set(1.0f, 1.0f);
+
+        uint16_t indices[6] = {0, 1, 2, 1, 2, 3};
+
+        fVtxBuffer = pipeline->fDevice.fMetalDevice->newBuffer(&vertexBuffer, sizeof(plateVertexBuffer), MTL::StorageModeManaged);
```

Comment:
I'm seeing this anti-pattern quite a bit. It feels like we should have a helper on `plMetalDevice` to make this less bugprone, such as:

```C++
template<typename T>
metalReturnType NewBuffer(const T& input, someMetalModeThingHere foo)
{
    return fMetalDevice->newBuffer(&input, sizeof(T));
}
```

This should work with any trivial complete type, so you can do:
```C++
uint16_t indices[6] = {0, 1, 2, 1, 2, 3};
idxBuffer = fDevice.NewBuffer(indices)
```
just fine.

You'd probably need something else for large heap allocated buffers or maybe a specialization for `std::unique_ptr<T[]>`, but I think that less manual byte counting is always a win.

Reply:
What worries me a bit is that MTL::Device is a pretty wide ranging API - and if we begin mirroring our device against it our API is going to get pretty bulky. I would follow simplification to a point (there should be an easy function to get the MTL::Device without going through layers of grossness.) But I'm not sure the device itself needs to be abstracted away.

It also helps that the returned buffer will be referenced managed by Metal itself with  - so we don't need to track it internally. (There is an option to turn off Metal managing buffer lifecycles - but I'm not inclined to do that right now.)

Reply:
To recap the discussion on IRC, what I'm worried about is the double calculation of the buffer size. That's vulnerable to accidental bugs. I'd prefer to see the compiler calculate the length of the buffer for us.

Reply:
Here's my thoughts:

I'm concerned about the definition of byte counting as an unsafe pattern. It is a dangerous pattern - but it's also extremely common in modern GPU programming.

- There are constant arrays like this one that need to go to the GPU - where the size is known at compile time.
- There are heap buffers that are dynamic that need to go to the GPU.
- Buffers are sometimes partially updated, requiring computation of an offset and a partial length.
- Shaders sometimes need to access buffers by offset and a sublength
- Sometimes shaders send buffers back to the GPU - which requires the CPU to infer length (common in compute shaders, lots of buffer math in the compute based character animation branch.)

What concerns me here is a bifurcation of the API by defining a function that is only useful for the first case, and cannot be used, or may even be dangerous for the other cases. I think that API will actually cause more confusion because it can only be used in very specific scenarios. I also think understanding byte math is necessary to work in Metal - and is extremely common in other Metal source bases. And byte math is done across both the CPU code and the shaders.

Even though I am unsure of this specific recommendation, there are still some things that I think could be good general improvements, and I can take a pass through the code:
- Buffer lengths should only be defined once (i.e. the CPU and GPU buffer allocations should not have separate lengths computed unless it's algorithmically necessary.)
- For constant arrays, `sizeof * count` should be avoided (which should be corrected in this block).
- It may be helpful to add utility functions for computing length, offsets, and subregions based on type. Narrowing the solution to the byte math itself would provide a common path for dynamic and constant arrays. In since MSL can import C++ code - these utilities could also be used in the shaders.
- _If this were Swift_ - Swift would let me modify the MTLDevice API and add convenience functions to it. And pointers in Swift can be annotated with size.

Reply:
> I'm concerned about the definition of byte counting as an unsafe pattern. It is a dangerous pattern - but it's also extremely common in modern GPU programming.

It's not that counting bytes in general is bad. It's when we count bytes once, then count the same bytes again a few lines later. This means we have two copies of the same calculation that need to be kept in sync. These kinds of anti-patterns are an easy source of bugs. I feel fairly strongly that there should be no double calculations at all.

Reply:
I think we're in sync then. I'll commit some changes here. I'll do a recheck as best I can based on the other parts of the code.

I think in the future I'll add some range math support so both shader code and CPU code have a more defined mechanism for calculating buffer manipulations. That should leave less room for error.

Reply:
Sounds like a good plan 👍

Reply:
Ok, I just committed 2e01bf91588dd0361781b70c5728f46331639d0c. It's probably not all the references, but a few things I'm did:

- I brought back inlCopy and friends from the D3D pipeline. Not all buffer copies fit with these functions (and the Direct3D pipeline doesn't use those functions in those cases either.) Long term - I think these should be moved into a header and shared. But it might be best to not drag D3D into this PR.
- Some sections of code were still manually computing sizes of vector types (e.g `sizeof(float) * 4` vs `sizeof(hsPoint4)`.) A lot of sections of code were always using sizeof against the vector types, but I found quite a few cases that didn't.
- In some cases I'm querying Apple's simd types instead of Headspin types for size. Everything _should_ all the be same size, so it probably doesn't matter. But it feels nicer to check simd sizes when I'm working with simd types in Metal, and hs types when the data comes from the engine.
- If the sizeof call is trivial (i.e. `sizeof(float)`) I'm not focusing on de-duplicating those. I am trying to de-duplicate sizeof calls where the result is complex or dynamic (i.e. `sizeof(float) * numFloatsInBuffer`)